File: repowire/peer_mcp.py
"""Per-peer MCP server config: list/add/remove across backends.

Same-host only for v1 — cross-host requires an ACP transport (not yet wired).
The daemon route layer enforces locality (peer.machine == this host) and circle
disambiguation; this module is pure backend-dispatch and is safe to call from
tests with mocked subprocesses or tmp_path-rooted config files.

Each backend exposes:

  list_servers(peer)         -> list[McpServerEntry]
  add_server(peer, spec)     -> None      (raises DuplicateServerError / NotSupportedError)
  remove_server(peer, name)  -> None      (raises ServerNotFoundError / NotSupportedError)

Errors are normalized into a small exception hierarchy so the route layer can
map them to consistent HTTP status codes (404, 409, 501, 504).
"""

from __future__ import annotations

import json
import os
import re
import shutil
import subprocess
import tempfile
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from repowire.agent_backends import agent_backend_for
from repowire.protocol.peers import Peer

# Subprocess timeout for `claude mcp` calls. Surfaces as 504 in the route.
SUBPROCESS_TIMEOUT_SECONDS = 10


# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------


@dataclass
class McpServerEntry:
    """One configured MCP server, normalized across backends."""

    name: str
    scope: str = "user"  # user | project (claude-code distinguishes)
    type: str = "stdio"  # stdio | http | sse
    command: str | None = None
    args: list[str] = field(default_factory=list)
    url: str | None = None
    env_keys: list[str] = field(default_factory=list)  # keys only, values redacted

    def to_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "scope": self.scope,
            "type": self.type,
            "command": self.command,
            "args": list(self.args),
            "url": self.url,
            "env_keys": list(self.env_keys),
        }


@dataclass
class McpServerSpec:
    """Input shape for add_server. Pass-through, no schema validation."""

    name: str
    type: str = "stdio"
    command: str | None = None
    args: list[str] = field(default_factory=list)
    url: str | None = None
    env: dict[str, str] = field(default_factory=dict)
    scope: str = "user"


# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------


class PeerMcpError(Exception):
    """Base error for peer MCP operations."""


class NotSupportedError(PeerMcpError):
    """Backend does not support MCP configuration in v1 (e.g. pi)."""


class DuplicateServerError(PeerMcpError):
    """Server with this name already exists."""


class ServerNotFoundError(PeerMcpError):
    """Named server is not configured."""


class BackendTimeoutError(PeerMcpError):
    """Backend CLI call timed out."""


class BackendError(PeerMcpError):
    """Backend CLI returned non-zero or unparseable output."""


class ValidationError(PeerMcpError):
    """MCP server input is unsafe or malformed."""


# ---------------------------------------------------------------------------
# Dispatch
# ---------------------------------------------------------------------------


_SERVER_NAME_RE = re.compile(r"^[A-Za-z0-9_-]{1,128}$")
_ENV_KEY_RE = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")
_RESERVED_SERVER_NAMES = {"__proto__", "constructor", "prototype"}


def _validate_server_name(name: str) -> None:
    if not name or not name.strip():
        raise ValidationError("server name is required")
    if name in _RESERVED_SERVER_NAMES:
        raise ValidationError(f"server name {name!r} is reserved")
    if not _SERVER_NAME_RE.fullmatch(name):
        raise ValidationError(
            "server name must contain only letters, numbers, underscores, or hyphens"
        )


def _validate_env_keys(env: dict[str, str]) -> None:
    for key in env:
        if not _ENV_KEY_RE.fullmatch(key):
            raise ValidationError(
                "env keys must start with a letter or underscore and contain only "
                "letters, numbers, or underscores"
            )


def _atomic_write_text(path: Path, text: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    mode = path.stat().st_mode & 0o777 if path.exists() else 0o600
    tmp_path: Path | None = None
    try:
        with tempfile.NamedTemporaryFile(
            "w",
            encoding="utf-8",
            dir=path.parent,
            prefix=f".{path.name}.",
            suffix=".tmp",
            delete=False,
        ) as tmp:
            tmp_path = Path(tmp.name)
            tmp.write(text)
            tmp.flush()
            os.fsync(tmp.fileno())
        tmp_path.chmod(mode)
        os.replace(tmp_path, path)
        try:
            dir_fd = os.open(path.parent, os.O_DIRECTORY)
        except OSError:
            return
        try:
            os.fsync(dir_fd)
        finally:
            os.close(dir_fd)
    finally:
        if tmp_path is not None and tmp_path.exists():
            tmp_path.unlink()


def list_servers(peer: Peer) -> list[McpServerEntry]:
    return agent_backend_for(peer.backend).list_mcp_servers(peer)


def add_server(peer: Peer, spec: McpServerSpec) -> None:
    _validate_server_name(spec.name)
    _validate_env_keys(spec.env)
    existing = {s.name for s in list_servers(peer)}
    if spec.name in existing:
        raise DuplicateServerError(f"server {spec.name!r} already exists")

    agent_backend_for(peer.backend).add_mcp_server(peer, spec)


def remove_server(peer: Peer, name: str) -> None:
    _validate_server_name(name)
    existing = {s.name for s in list_servers(peer)}
    if name not in existing:
        raise ServerNotFoundError(f"server {name!r} not configured")

    agent_backend_for(peer.backend).remove_mcp_server(peer, name)


# ---------------------------------------------------------------------------
# claude-code: shells out to the `claude mcp` CLI
# ---------------------------------------------------------------------------


def _claude_cli() -> str:
    path = shutil.which("claude")
    if not path:
        raise NotSupportedError("`claude` CLI not found on PATH")
    return path


def _run_claude(args: list[str], cwd: str | None = None) -> subprocess.CompletedProcess:
    cli = _claude_cli()
    try:
        return subprocess.run(
            [cli, *args],
            cwd=cwd,
            capture_output=True,
            text=True,
            timeout=SUBPROCESS_TIMEOUT_SECONDS,
            check=False,
        )
    except subprocess.TimeoutExpired as e:
        raise BackendTimeoutError(f"`claude {' '.join(args)}` timed out") from e


def _claude_list(peer: Peer) -> list[McpServerEntry]:
    """Use `claude mcp list` (text output). JSON flag isn't stable across
    versions, so we parse the human-readable list — one line per server like:

        servername: command arg1 arg2 - ✓ Connected
        urlname: https://example.com (SSE) - ✓ Connected
    """
    result = _run_claude(["mcp", "list"], cwd=peer.path or None)
    if result.returncode != 0:
        # CLI exits non-zero when no servers configured on some versions
        stdout_lower = result.stdout.lower()
        stderr_lower = result.stderr.lower()
        if "no mcp servers" in stdout_lower or "no mcp servers" in stderr_lower:
            return []
        detail = result.stderr.strip() or result.stdout.strip()
        raise BackendError(f"claude mcp list failed: {detail}")

    out: list[McpServerEntry] = []
    for line in result.stdout.splitlines():
        line = line.strip()
        if not line or ":" not in line:
            continue
        # Skip header/footer lines that don't have the "name: detail" shape
        name, _, rest = line.partition(":")
        name = name.strip()
        if not name or " " in name:
            continue
        rest = rest.strip()
        # Strip trailing status decorations like " - ✓ Connected"
        rest = re.sub(r"\s*-\s*[✓✗].*$", "", rest).strip()
        if rest.startswith(("http://", "https://")):
            url = rest.split()[0]
            entry = McpServerEntry(name=name, type="http", url=url)
        else:
            parts = rest.split()
            command = parts[0] if parts else None
            args = parts[1:] if len(parts) > 1 else []
            entry = McpServerEntry(name=name, type="stdio", command=command, args=args)
        out.append(entry)
    return out


def _claude_add(peer: Peer, spec: McpServerSpec) -> None:
    scope_flag = ["-s", "project"] if spec.scope == "project" else ["-s", "user"]
    env_flags: list[str] = []
    for k, v in spec.env.items():
        env_flags.extend(["-e", f"{k}={v}"])

    if spec.type in ("http", "sse"):
        if not spec.url:
            raise BackendError("url is required for http/sse type")
        transport = ["--transport", spec.type]
        args = ["mcp", "add", *scope_flag, *transport, *env_flags, spec.name, spec.url]
    else:
        if not spec.command:
            raise BackendError("command is required for stdio type")
        args = ["mcp", "add", *scope_flag, *env_flags, spec.name, "--", spec.command, *spec.args]

    cwd = peer.path if spec.scope == "project" else None
    result = _run_claude(args, cwd=cwd)
    if result.returncode != 0:
        detail = result.stderr.strip() or result.stdout.strip()
        raise BackendError(f"claude mcp add failed: {detail}")


def _claude_remove(peer: Peer, name: str) -> None:
    result = _run_claude(["mcp", "remove", name], cwd=peer.path or None)
    if result.returncode != 0:
        detail = result.stderr.strip() or result.stdout.strip()
        raise BackendError(f"claude mcp remove failed: {detail}")


# ---------------------------------------------------------------------------
# codex: edit ~/.codex/config.toml
# ---------------------------------------------------------------------------

CODEX_CONFIG_PATH = Path.home() / ".codex" / "config.toml"


_SECTION_RE = re.compile(r"^\[mcp_servers\.([^\]]+)\]\s*$")
_KV_RE = re.compile(r'^(\w+)\s*=\s*(.+?)\s*$')


def _parse_toml_value(raw: str) -> Any:
    """Minimal TOML value parser: handles strings, lists of strings, inline tables."""
    raw = raw.strip()
    if raw.startswith('"') and raw.endswith('"'):
        return raw[1:-1]
    if raw.startswith("[") and raw.endswith("]"):
        inner = raw[1:-1].strip()
        if not inner:
            return []
        items = re.findall(r'"([^"]*)"', inner)
        return items
    if raw.startswith("{") and raw.endswith("}"):
        out: dict[str, str] = {}
        for pair in re.findall(r'(\w+)\s*=\s*"([^"]*)"', raw):
            out[pair[0]] = pair[1]
        return out
    return raw


def _codex_read_sections() -> dict[str, dict[str, Any]]:
    """Parse only `[mcp_servers.NAME]` sections. Returns name -> {key: parsed_value}."""
    if not CODEX_CONFIG_PATH.exists():
        return {}
    sections: dict[str, dict[str, Any]] = {}
    current: str | None = None
    for line in CODEX_CONFIG_PATH.read_text().splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            continue
        m = _SECTION_RE.match(stripped)
        if m:
            current = m.group(1)
            sections[current] = {}
            continue
        if stripped.startswith("["):
            current = None  # entered some other section
            continue
        if current is None:
            continue
        kv = _KV_RE.match(stripped)
        if kv:
            sections[current][kv.group(1)] = _parse_toml_value(kv.group(2))
    return sections


def _codex_list() -> list[McpServerEntry]:
    out: list[McpServerEntry] = []
    for name, body in _codex_read_sections().items():
        env = body.get("env", {})
        env_keys: list[str] = [str(k) for k in env.keys()] if isinstance(env, dict) else []
        cmd = body.get("command")
        args = body.get("args", [])
        if not isinstance(args, list):
            args = []
        url = body.get("url")
        srv_type = "http" if url else "stdio"
        out.append(
            McpServerEntry(
                name=name,
                scope="user",
                type=srv_type,
                command=cmd if isinstance(cmd, str) else None,
                args=[str(a) for a in args],
                url=url if isinstance(url, str) else None,
                env_keys=env_keys,
            )
        )
    return out


def _toml_quote(s: str) -> str:
    return json.dumps(s)


def _codex_render_section(spec: McpServerSpec) -> str:
    lines = [f"[mcp_servers.{spec.name}]"]
    if spec.command:
        lines.append(f"command = {_toml_quote(spec.command)}")
    if spec.args:
        rendered_args = ", ".join(_toml_quote(a) for a in spec.args)
        lines.append(f"args = [{rendered_args}]")
    if spec.url:
        lines.append(f"url = {_toml_quote(spec.url)}")
    if spec.env:
        rendered_env = ", ".join(f"{k} = {_toml_quote(v)}" for k, v in spec.env.items())
        lines.append(f"env = {{ {rendered_env} }}")
    return "\n".join(lines) + "\n"


def _codex_add(spec: McpServerSpec) -> None:
    section = _codex_render_section(spec)
    if CODEX_CONFIG_PATH.exists():
        existing = CODEX_CONFIG_PATH.read_text()
        sep = "" if existing.endswith("\n\n") else ("\n" if existing.endswith("\n") else "\n\n")
        _atomic_write_text(CODEX_CONFIG_PATH, existing + sep + section)
    else:
        _atomic_write_text(CODEX_CONFIG_PATH, section)


def _codex_remove(name: str) -> None:
    if not CODEX_CONFIG_PATH.exists():
        raise ServerNotFoundError(f"server {name!r} not configured")
    lines = CODEX_CONFIG_PATH.read_text().splitlines(keepends=True)
    out: list[str] = []
    skipping = False
    target = f"[mcp_servers.{name}]"
    for line in lines:
        stripped = line.strip()
        if stripped == target:
            skipping = True
            continue
        if skipping and stripped.startswith("[") and stripped.endswith("]"):
            skipping = False
        if not skipping:
            out.append(line)
    _atomic_write_text(CODEX_CONFIG_PATH, "".join(out))


# ---------------------------------------------------------------------------
# gemini: edit ~/.gemini/settings.json mcpServers block
# ---------------------------------------------------------------------------

GEMINI_SETTINGS_PATH = Path.home() / ".gemini" / "settings.json"


def _gemini_load() -> dict[str, Any]:
    if not GEMINI_SETTINGS_PATH.exists():
        return {}
    try:
        return json.loads(GEMINI_SETTINGS_PATH.read_text())
    except (json.JSONDecodeError, OSError) as e:
        raise BackendError(f"failed to read gemini settings: {e}") from e


def _gemini_save(data: dict[str, Any]) -> None:
    _atomic_write_text(GEMINI_SETTINGS_PATH, json.dumps(data, indent=2))


def _gemini_list() -> list[McpServerEntry]:
    data = _gemini_load()
    servers = data.get("mcpServers", {})
    if not isinstance(servers, dict):
        return []
    out: list[McpServerEntry] = []
    for name, body in servers.items():
        if not isinstance(body, dict):
            continue
        env = body.get("env", {})
        env_keys: list[str] = [str(k) for k in env.keys()] if isinstance(env, dict) else []
        url = body.get("url") or body.get("httpUrl")
        srv_type = "http" if url else "stdio"
        out.append(
            McpServerEntry(
                name=name,
                scope="user",
                type=srv_type,
                command=body.get("command") if isinstance(body.get("command"), str) else None,
                args=list(body.get("args", []) or []),
                url=url if isinstance(url, str) else None,
                env_keys=env_keys,
            )
        )
    return out


def _gemini_add(spec: McpServerSpec) -> None:
    data = _gemini_load()
    servers = data.get("mcpServers", {})
    if not isinstance(servers, dict):
        raise BackendError("gemini settings mcpServers must be an object")
    data["mcpServers"] = servers
    entry: dict[str, Any] = {}
    if spec.command:
        entry["command"] = spec.command
    if spec.args:
        entry["args"] = list(spec.args)
    if spec.url:
        entry["url"] = spec.url
    if spec.env:
        entry["env"] = dict(spec.env)
    servers[spec.name] = entry
    _gemini_save(data)


def _gemini_remove(name: str) -> None:
    data = _gemini_load()
    servers = data.get("mcpServers", {})
    if not isinstance(servers, dict) or name not in servers:
        raise ServerNotFoundError(f"server {name!r} not configured")
    del servers[name]
    _gemini_save(data)

File: daemon-go/service/peer_mcp.go
package service

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/repowire/repowire/daemon-go/proto"
)

type MCPServerEntry struct {
	Name    string   `json:"name"`
	Scope   string   `json:"scope"`
	Type    string   `json:"type"`
	Command *string  `json:"command"`
	Args    []string `json:"args"`
	URL     *string  `json:"url"`
	EnvKeys []string `json:"env_keys"`
}

type MCPServerSpec struct {
	Name    string            `json:"name"`
	Type    string            `json:"type"`
	Command *string           `json:"command,omitempty"`
	Args    []string          `json:"args,omitempty"`
	URL     *string           `json:"url,omitempty"`
	Env     map[string]string `json:"env,omitempty"`
	Scope   string            `json:"-"`
}

var (
	ErrMCPUnsupported = errors.New("MCP config not supported for backend")
	ErrMCPDuplicate   = errors.New("MCP server already exists")
	ErrMCPNotFound    = errors.New("MCP server not configured")
	serverNameRE      = regexp.MustCompile(`^[A-Za-z0-9_-]{1,128}$`)
	envKeyRE          = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)
	codexSectionRE    = regexp.MustCompile(`^\[mcp_servers\.([^]]+)\]$`)
)

func validateMCPSpec(spec MCPServerSpec) error {
	if !serverNameRE.MatchString(spec.Name) || spec.Name == "__proto__" || spec.Name == "constructor" || spec.Name == "prototype" {
		return errors.New("server name must contain only letters, numbers, underscores, or hyphens")
	}
	for key := range spec.Env {
		if !envKeyRE.MatchString(key) {
			return errors.New("env keys must start with a letter or underscore and contain only letters, numbers, or underscores")
		}
	}
	if spec.Type == "" {
		spec.Type = "stdio"
	}
	if spec.Type == "stdio" && (spec.Command == nil || *spec.Command == "") {
		return errors.New("command is required for stdio type")
	}
	if (spec.Type == "http" || spec.Type == "sse") && (spec.URL == nil || *spec.URL == "") {
		return errors.New("url is required for http/sse type")
	}
	if spec.Type != "stdio" && spec.Type != "http" && spec.Type != "sse" {
		return errors.New("type must be stdio, http, or sse")
	}
	return nil
}

func MCPConfigScope(backend proto.AgentType) (map[string]any, error) {
	base := map[string]any{"backend": backend, "default_scope": "user"}
	switch backend {
	case proto.AgentClaudeCode:
		base["owner"], base["effective_scope"], base["label"] = "peer/project", "peer_project", "Claude Code peer/project config"
		base["description"], base["supported_scopes"], base["is_global"] = "Claude Code MCP edits can target user/global config or the peer's project/worktree via the selected add scope.", []string{"user", "project"}, false
	case proto.AgentCodex:
		base["owner"], base["effective_scope"], base["label"] = "backend", "backend_global", "Codex global backend config"
		base["description"], base["supported_scopes"], base["is_global"] = "Codex MCP edits target the user-level Codex config shared by Codex sessions on this host.", []string{"user"}, true
	default:
		return nil, fmt.Errorf("%w %s", ErrMCPUnsupported, backend)
	}
	return base, nil
}

func ListPeerMCP(ctx context.Context, peer *proto.Peer) ([]MCPServerEntry, error) {
	switch peer.Backend {
	case proto.AgentClaudeCode:
		return listClaudeMCP(ctx, peer.Path)
	case proto.AgentCodex:
		return listCodexMCP()
	default:
		return nil, fmt.Errorf("%w %s", ErrMCPUnsupported, peer.Backend)
	}
}

func AddPeerMCP(ctx context.Context, peer *proto.Peer, spec MCPServerSpec) error {
	if spec.Type == "" {
		spec.Type = "stdio"
	}
	if err := validateMCPSpec(spec); err != nil {
		return err
	}
	entries, err := ListPeerMCP(ctx, peer)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		if entry.Name == spec.Name {
			return fmt.Errorf("%w: %q", ErrMCPDuplicate, spec.Name)
		}
	}
	switch peer.Backend {
	case proto.AgentClaudeCode:
		return addClaudeMCP(ctx, peer.Path, spec)
	case proto.AgentCodex:
		return addCodexMCP(spec)
	default:
		return fmt.Errorf("%w %s", ErrMCPUnsupported, peer.Backend)
	}
}

func RemovePeerMCP(ctx context.Context, peer *proto.Peer, name string) error {
	if !serverNameRE.MatchString(name) {
		return errors.New("invalid server name")
	}
	entries, err := ListPeerMCP(ctx, peer)
	if err != nil {
		return err
	}
	found := false
	for _, entry := range entries {
		if entry.Name == name {
			found = true
			break
		}
	}
	if !found {
		return fmt.Errorf("%w: %q", ErrMCPNotFound, name)
	}
	switch peer.Backend {
	case proto.AgentClaudeCode:
		return runClaudeMCP(ctx, peer.Path, "mcp", "remove", name)
	case proto.AgentCodex:
		return removeCodexMCP(name)
	default:
		return fmt.Errorf("%w %s", ErrMCPUnsupported, peer.Backend)
	}
}

func runClaudeMCP(ctx context.Context, cwd string, args ...string) error {
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, "claude", args...)
	if cwd != "" {
		cmd.Dir = cwd
	}
	out, err := cmd.CombinedOutput()
	if ctx.Err() != nil {
		return fmt.Errorf("claude %s timed out", strings.Join(args, " "))
	}
	if err != nil {
		return fmt.Errorf("claude %s failed: %s", strings.Join(args, " "), strings.TrimSpace(string(out)))
	}
	return nil
}

func listClaudeMCP(ctx context.Context, cwd string) ([]MCPServerEntry, error) {
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, "claude", "mcp", "list")
	if cwd != "" {
		cmd.Dir = cwd
	}
	out, err := cmd.CombinedOutput()
	if ctx.Err() != nil {
		return nil, errors.New("claude mcp list timed out")
	}
	if err != nil {
		if strings.Contains(strings.ToLower(string(out)), "no mcp servers") {
			return []MCPServerEntry{}, nil
		}
		return nil, fmt.Errorf("claude mcp list failed: %s", strings.TrimSpace(string(out)))
	}
	entries := []MCPServerEntry{}
	for _, line := range strings.Split(string(out), "\n") {
		name, rest, ok := strings.Cut(strings.TrimSpace(line), ":")
		if !ok || name == "" || strings.Contains(name, " ") {
			continue
		}
		if index := strings.Index(rest, " - "); index >= 0 {
			rest = rest[:index]
		}
		fields := strings.Fields(strings.TrimSpace(rest))
		if len(fields) == 0 {
			continue
		}
		entry := MCPServerEntry{Name: name, Scope: "user", Type: "stdio", Args: []string{}, EnvKeys: []string{}}
		if strings.HasPrefix(fields[0], "http://") || strings.HasPrefix(fields[0], "https://") {
			entry.Type = "http"
			entry.URL = &fields[0]
		} else {
			entry.Command = &fields[0]
			entry.Args = fields[1:]
		}
		entries = append(entries, entry)
	}
	return entries, nil
}

func addClaudeMCP(ctx context.Context, cwd string, spec MCPServerSpec) error {
	scope := spec.Scope
	if scope == "" {
		scope = "user"
	}
	args := []string{"mcp", "add", "-s", scope}
	if spec.Type == "http" || spec.Type == "sse" {
		args = append(args, "--transport", spec.Type)
	}
	keys := make([]string, 0, len(spec.Env))
	for key := range spec.Env {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		args = append(args, "-e", key+"="+spec.Env[key])
	}
	args = append(args, spec.Name)
	if spec.Type == "stdio" {
		args = append(args, "--", *spec.Command)
		args = append(args, spec.Args...)
	} else {
		args = append(args, *spec.URL)
	}
	if scope != "project" {
		cwd = ""
	}
	return runClaudeMCP(ctx, cwd, args...)
}

func codexMCPPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".codex", "config.toml")
}
func listCodexMCP() ([]MCPServerEntry, error) {
	raw, err := os.ReadFile(codexMCPPath())
	if os.IsNotExist(err) {
		return []MCPServerEntry{}, nil
	}
	if err != nil {
		return nil, err
	}
	sections := parseCodexSections(string(raw))
	names := make([]string, 0, len(sections))
	for name := range sections {
		names = append(names, name)
	}
	sort.Strings(names)
	entries := []MCPServerEntry{}
	for _, name := range names {
		body := sections[name]
		entry := MCPServerEntry{Name: name, Scope: "user", Type: "stdio", Args: parseTOMLStrings(body["args"]), EnvKeys: parseTOMLEnvKeys(body["env"])}
		if value := parseTOMLString(body["command"]); value != "" {
			entry.Command = &value
		}
		if value := parseTOMLString(body["url"]); value != "" {
			entry.URL = &value
			entry.Type = "http"
		}
		entries = append(entries, entry)
	}
	return entries, nil
}
func parseCodexSections(content string) map[string]map[string]string {
	out := map[string]map[string]string{}
	current := ""
	for _, line := range strings.Split(content, "\n") {
		trim := strings.TrimSpace(line)
		if match := codexSectionRE.FindStringSubmatch(trim); match != nil {
			current = match[1]
			out[current] = map[string]string{}
			continue
		}
		if strings.HasPrefix(trim, "[") {
			current = ""
			continue
		}
		if current != "" {
			if key, value, ok := strings.Cut(trim, "="); ok {
				out[current][strings.TrimSpace(key)] = strings.TrimSpace(value)
			}
		}
	}
	return out
}
func parseTOMLString(raw string) string { value, _ := strconv.Unquote(raw); return value }
func parseTOMLStrings(raw string) []string {
	var out []string
	_ = json.Unmarshal([]byte(raw), &out)
	if out == nil {
		out = []string{}
	}
	return out
}
func parseTOMLEnvKeys(raw string) []string {
	matches := regexp.MustCompile(`([A-Za-z_][A-Za-z0-9_]*)\s*=`).FindAllStringSubmatch(raw, -1)
	out := make([]string, 0, len(matches))
	for _, match := range matches {
		out = append(out, match[1])
	}
	sort.Strings(out)
	return out
}
func addCodexMCP(spec MCPServerSpec) error {
	var body strings.Builder
	fmt.Fprintf(&body, "[mcp_servers.%s]\n", spec.Name)
	if spec.Command != nil {
		raw, _ := json.Marshal(*spec.Command)
		fmt.Fprintf(&body, "command = %s\n", raw)
	}
	if len(spec.Args) > 0 {
		raw, _ := json.Marshal(spec.Args)
		fmt.Fprintf(&body, "args = %s\n", raw)
	}
	if spec.URL != nil {
		raw, _ := json.Marshal(*spec.URL)
		fmt.Fprintf(&body, "url = %s\n", raw)
	}
	if len(spec.Env) > 0 {
		keys := make([]string, 0, len(spec.Env))
		for key := range spec.Env {
			keys = append(keys, key)
		}
		sort.Strings(keys)
		pairs := []string{}
		for _, key := range keys {
			raw, _ := json.Marshal(spec.Env[key])
			pairs = append(pairs, key+" = "+string(raw))
		}
		fmt.Fprintf(&body, "env = { %s }\n", strings.Join(pairs, ", "))
	}
	existing, err := os.ReadFile(codexMCPPath())
	if err != nil && !os.IsNotExist(err) {
		return err
	}
	content := strings.TrimRight(string(existing), "\n")
	if content != "" {
		content += "\n\n"
	}
	content += body.String()
	return atomicWrite(codexMCPPath(), []byte(content), 0o600)
}
func removeCodexMCP(name string) error {
	path := codexMCPPath()
	raw, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	var out []string
	skipping := false
	for _, line := range strings.SplitAfter(string(raw), "\n") {
		trim := strings.TrimSpace(line)
		if trim == "[mcp_servers."+name+"]" {
			skipping = true
			continue
		}
		if skipping && strings.HasPrefix(trim, "[") {
			skipping = false
		}
		if !skipping {
			out = append(out, line)
		}
	}
	return atomicWrite(path, []byte(strings.Join(out, "")), 0o600)
}

func atomicWrite(path string, content []byte, mode os.FileMode) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}
	tmp, err := os.CreateTemp(filepath.Dir(path), "."+filepath.Base(path)+".*.tmp")
	if err != nil {
		return err
	}
	tmpPath := tmp.Name()
	defer os.Remove(tmpPath)
	if _, err := bytes.NewReader(content).WriteTo(tmp); err != nil {
		tmp.Close()
		return err
	}
	if err := tmp.Sync(); err != nil {
		tmp.Close()
		return err
	}
	if err := tmp.Close(); err != nil {
		return err
	}
	if err := os.Chmod(tmpPath, mode); err != nil {
		return err
	}
	return os.Rename(tmpPath, path)
}
